SonarQube.MSBuild.Tasks/MakeUniqueDir.cs

```diff
@@ -0,0 +1,101 @@
+﻿/*
+ * SonarQube Scanner for MSBuild
+ * Copyright (C) 2016-2017 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+
+using System;
+using System.IO;
+using System.Text;
+using Microsoft.Build.Framework;
+using Microsoft.Build.Utilities;
+
+using Mutex = System.Threading.Mutex;
+
+namespace SonarQube.MSBuild.Tasks
+{
+    public class MakeUniqueDir : Task
+    {
+        /// <summary>
+        /// The path where folder will be created, usually that is .sonarqube/out in the solution root
+        /// </summary>
+        [Required]
+        public string Path { get; set; }
+
+        /// <summary>
+        /// Optional folder name prefix
+        /// </summary>
+        public string Prefix { get; set; }
+
+        /// <summary>
+        /// Returns the name of the created folder
+        /// </summary>
+        [Output]
+        public string UniqueName { get; set; }
+
+        /// <summary>
+        /// Returns the full path of the created folder
+        /// </summary>
+        [Output]
+        public string UniquePath { get; set; }
+
+        public override bool Execute()
+        {
+            var mutex = new Mutex(false, StripReservedChars(Path));
```

Comment:
[might be dumb] As this class is not a singleton, can't we have this call being triggered in parallel and so ending up with the same file name?

Reply:
It is fine if it is triggered in parallel - the next line will be synchronized and will wait for the previous call to release